torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+      TORCH_INTERNAL_ASSERT(n->blocks().size() == 1);
+      convertReturnsToTuples(n->blocks()[0]);
+
+      Graph* g = b->owningGraph();
+      Block* sub_block = n->blocks()[0];
+      if (sub_block->outputs().size() > 1) {
+        {
+          // Make block returns go through a Tuple
+          WithInsertPoint guard(sub_block->return_node());
+          Node* return_tup =
+              g->insertNode(g->createTuple(sub_block->outputs()));
+          while (sub_block->outputs().size()) {
+            sub_block->eraseOutput(0);
+          }
+          sub_block->registerOutput(return_tup->output());
+        }
+
+        // Make node outputs a single tuple;
+        std::vector<TypePtr> types;
+        for (size_t i = 0; i < n->outputs().size(); ++i) {
+          types.push_back(n->output(0)->type());
+        }
+        Value* tup_output = n->addOutput()->setType(TupleType::create(types));
+        Node* tup_unpack = g->createTupleUnpack(tup_output)->insertAfter(n);
+        for (size_t i = 0; i < tup_unpack->outputs().size(); ++i) {
+          n->output(0)->replaceAllUsesWith(tup_unpack->output(i));
+          n->eraseOutput(0);
+        }
+      } else if (sub_block->outputs().size() == 0) {
+        WithInsertPoint guard(sub_block->return_node());
+        sub_block->registerOutput(g->insertNode(g->createNone())->output());
+        n->addOutput()->setType(NoneType::get());
+      }
+    }
+  }
+}
+
+void lambdaLiftBlocksAndConvertToGraph(Block* b) {
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      lambdaLiftBlocksAndConvertToGraph(n->blocks()[0]);
+
+      auto graph = std::make_shared<Graph>();
+      std::unordered_map<Value*, Value*> remaps;
+      graph->block()->cloneFrom(n->blocks()[0], [&](Value* v) {
+        if (!remaps.count(v)) {
+          remaps[v] = graph->addInput()->copyMetadata(v);
+          n->addInput(v);
+        }
+        return remaps[v];
+      });
+      LintGraph(graph);
+      n->g_(attr::Subgraph, graph);
+      n->eraseBlock(0);
+    }
+  }
+
+  // Inline base __module call
+  for (auto n_itr = b->nodes().begin(); n_itr != b->nodes().end();) {
+    Node* n = *n_itr++;
+    if (n->kind() == prim::FakeScopeBlock && n->s(attr::scope) == "__module") {
+      inlineCallTo(n, *n->g(attr::Subgraph));
+    }
+  }
+}
+
+namespace {
+
+std::vector<std::string> splitModuleQualname(const std::string& s) {
+  std::vector<std::string> retval;
+  size_t start = 0;
+  size_t end = s.find('.');
+  while (end != std::string::npos) {
+    retval.push_back(s.substr(start, end - start));
+    start = end + 1;
+    end = s.find('.', start);
+  }
+
+  retval.push_back(s.substr(start, end));
+  return retval;
+}
+
+} // namespace
+
+void createMethodCalls(
+    const std::shared_ptr<Graph>& g,
+    script::Module this_level_self,
+    std::vector<std::string> prefix) {
+  Value* self;
+  if (g->inputs().size() == 0 ||
+      g->inputs()[0]->type() != this_level_self.type()) {
```

Comment:
I don't know what this is suppose to be checking for, but pointer comparison of the types is most likely wrong.

Reply:
Something feels wrong here. There is no guarantee that something with the same type as self is the self module (we should not assume they are singletons).  I feel like Module self should just be the first argument everywhere so it doesn't have to be recovered here.